### ISSUE TITLE
Some PerlIO layers can't be stacked during open

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11054]
Migrated from [rt.perl.org#82484](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=82484) (status was 'resolved')

Searchable as RT82484$

**Comment by p5pRT** [comment]
### From @Leont


This is a bug report for perl from fawaka@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.




---
Three of Perl's builtin PerlIO layers \(«&#8203;:pop»\, «&#8203;:utf8» and «&#8203;:bytes»\) do not allow stacking when opening a file\. For example this&#8203;:

 open FH\, '>&#8203;:pop&#8203;:perlio'\, 'some\.file' or die $\!;

Will throw an error&#8203;: «Invalid argument»\. There's no reason why this shouldn't be allowed\, and particular pop is sometimes useful\.



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.10\.1&#8203;:

Configured by Debian Project at Mon Jul 12 12&#8203;:19&#8203;:34 UTC 2010\.

Summary of my perl5 \(revision 5 version 10 subversion 1\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.24\-27\-server\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux yellow 2\.6\.24\-27\-server \#1 smp fri mar 12 01&#8203;:23&#8203;:09 utc 2010 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.10 \-Darchlib=/usr/lib/perl/5\.10 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.10\.1 \-Dsitearch=/usr/local/lib/perl/5\.10\.1 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.10\.1 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.4\.4'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.12\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.10\.1
&nbsp;    gnulibc\_version='2\.12'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/arm\_thread\_stress\_timeout \- http&#8203;://bugs\.debian\.org/501970 Raise the timeout of ext/threads/shared/t/stress\.t to accommodate slower build hosts
&nbsp;    DEBPKG&#8203;:debian/cpan\_config\_path \- Set location of CPAN&#8203;::Config to /etc/perl as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- http&#8203;://bugs\.debian\.org/340047 Remove overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- http&#8203;://bugs\.debian\.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/errno\_ver \- http&#8203;://bugs\.debian\.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
&nbsp;    DEBPKG&#8203;:debian/extutils\_hacks \- Various debian\-specific ExtUtils changes
&nbsp;    DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
&nbsp;    DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
&nbsp;    DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/m68k\_thread\_stress \- http&#8203;://bugs\.debian\.org/495826 Disable some threads tests on m68k for now due to missing TLS\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/module\_build\_man\_extensions \- http&#8203;://bugs\.debian\.org/479460 Adjust Module&#8203;::Build manual page extensions for the Debian Perl policy
&nbsp;    DEBPKG&#8203;:debian/perl\_synopsis \- http&#8203;://bugs\.debian\.org/278323 Rearrange perl\.pod
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- http&#8203;://bugs\.debian\.org/128355 Prune the list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:debian/use\_gdbm \- Explicitly link against \-lgdbm\_compat in ODBM\_File/NDBM\_File\. 
&nbsp;    DEBPKG&#8203;:fixes/assorted\_docs \- http&#8203;://bugs\.debian\.org/443733 \[384f06a\] Math&#8203;::BigInt&#8203;::CalcEmu documentation grammar fix
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- http&#8203;://bugs\.debian\.org/100195 \[rt\.cpan\.org \#36038\] Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:fixes/processPL \- http&#8203;://bugs\.debian\.org/357264 \[rt\.cpan\.org \#17224\] Always use PERLRUNINST when building perl modules\.
&nbsp;    DEBPKG&#8203;:debian/perlivp \- http&#8203;://bugs\.debian\.org/510895 Make perlivp skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:fixes/pod2man\-index\-backslash \- http&#8203;://bugs\.debian\.org/521256 Escape backslashes in \.IX entries
&nbsp;    DEBPKG&#8203;:debian/disable\-zlib\-bundling \- Disable zlib bundling in Compress&#8203;::Raw&#8203;::Zlib
&nbsp;    DEBPKG&#8203;:fixes/kfreebsd\_cppsymbols \- http&#8203;://bugs\.debian\.org/533098 \[3b910a0\] Add gcc predefined macros to $Config\{cppsymbols\} on GNU/kFreeBSD\.
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_definstalldirs \- http&#8203;://bugs\.debian\.org/533707 Configure CPANPLUS to use the site directories by default\.
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_config\_path \- Save local versions of CPANPLUS&#8203;::Config&#8203;::System into /etc/perl\.
&nbsp;    DEBPKG&#8203;:fixes/kfreebsd\-filecopy\-pipes \- http&#8203;://bugs\.debian\.org/537555 \[16f708c\] Fix File&#8203;::Copy&#8203;::copy with pipes on GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/anon\-tmpfile\-dir \- http&#8203;://bugs\.debian\.org/528544 \[perl \#66452\] Honor TMPDIR when open\(\)ing an anonymous temporary file
&nbsp;    DEBPKG&#8203;:fixes/abstract\-sockets \- http&#8203;://bugs\.debian\.org/329291 \[89904c0\] Add support for Abstract namespace sockets\.
&nbsp;    DEBPKG&#8203;:fixes/hurd\_cppsymbols \- http&#8203;://bugs\.debian\.org/544307 \[eeb92b7\] Add gcc predefined macros to $Config\{cppsymbols\} on GNU/Hurd\.
&nbsp;    DEBPKG&#8203;:fixes/autodie\-flock \- http&#8203;://bugs\.debian\.org/543731 Allow for flock returning EAGAIN instead of EWOULDBLOCK on linux/parisc
&nbsp;    DEBPKG&#8203;:fixes/archive\-tar\-instance\-error \- http&#8203;://bugs\.debian\.org/539355 \[rt\.cpan\.org \#48879\] Separate Archive&#8203;::Tar instance error strings from each other
&nbsp;    DEBPKG&#8203;:fixes/positive\-gpos \- http&#8203;://bugs\.debian\.org/545234 \[perl \#69056\] \[c584a96\] Fix \\\\G crash on first match
&nbsp;    DEBPKG&#8203;:debian/devel\-ppport\-ia64\-optim \- http&#8203;://bugs\.debian\.org/548943 Work around an ICE on ia64
&nbsp;    DEBPKG&#8203;:fixes/trie\-logic\-match \- http&#8203;://bugs\.debian\.org/552291 \[perl \#69973\] \[0abd0d7\] Fix a DoS in Unicode processing \[CVE\-2009\-3626\]
&nbsp;    DEBPKG&#8203;:fixes/hppa\-thread\-eagain \- http&#8203;://bugs\.debian\.org/554218 make the threads\-shared test suite more robust\, fixing failures on hppa
&nbsp;    DEBPKG&#8203;:fixes/crash\-on\-undefined\-destroy \- http&#8203;://bugs\.debian\.org/564074 \[perl \#71952\] \[1f15e67\] Fix a NULL pointer dereference when looking for a DESTROY method
&nbsp;    DEBPKG&#8203;:fixes/tainted\-errno \- http&#8203;://bugs\.debian\.org/574129 \[perl \#61976\] \[be1cf43\] fix an errno stringification bug in taint mode
&nbsp;    DEBPKG&#8203;:patchlevel \- http&#8203;://bugs\.debian\.org/567489 List packaged patches for 5\.10\.1\-12 in patchlevel\.h



---
@&#8203;INC for perl 5\.10\.1&#8203;:
&nbsp;    /home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/leon/perl5/lib/perl5
&nbsp;    /home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/leon/perl5/lib/perl5
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.10\.1
&nbsp;    /usr/local/share/perl/5\.10\.1
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.10
&nbsp;    /usr/share/perl/5\.10
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.10\.1&#8203;:
&nbsp;    HOME=/home/leon
&nbsp;    LANG=en\_US\.utf8
&nbsp;    LANGUAGE=en\_US&#8203;:en\_GB&#8203;:en
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/leon/perl5/perlbrew/bin&#8203;:/home/leon/perl5/perlbrew/perls/current/bin&#8203;:/home/leon/perl5/bin&#8203;:/home/leon/bin&#8203;:/home/leon/perl5/bin&#8203;:/home/leon/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/games
&nbsp;    PERL5LIB=/home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi&#8203;:/home/leon/perl5/lib/perl5&#8203;:/home/leon/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi&#8203;:/home/leon/perl5/lib/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_MM\_OPT=INSTALL\_BASE=/home/leon/perl5
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @Leont

The problem behind this can be summarized to this&#8203;: «during open\(\) a
PerlIO layer that defines an Open method can not be pushed on top of
another layer that doesn't define Open»\. There are two possible
workarounds for this&#8203;:

a\) Implement Open in all layers that don't implement it now\.
b\) Change the lookup mechanism in all existing Open methods\.

I decided to implement option a\, because it seem to be the least
intrusive way to tackle this problem\. In the first patch\, I introduce
PerlIOBase\_open\, a generic open method that does what 90%\+ of all
modules want it to do \(open the layer beneath it\, and then push the
current layer on top of it\)\. It also updates the documentation to
advise people to use it\.

The second patch used this to fix this bug \(\#82484\)\. The third patch
fixes \#80764 using the same Open method \(consider my previous patch
for that bug retracted in favor of this one\)\. These two patches are
quite trivial :\-\)\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>0001-Define-PerlIOBase_open.patch</summary>

```diff
From 262cdd9df41753283150a2a68d121300852138f3 Mon Sep 17 00:00:00 2001
From: Leon Timmermans <fawaka@gmail.com>
Date: Thu, 20 Jan 2011 21:08:34 +0100
Subject: [PATCH 1/3] Define PerlIOBase_open

---
 perlio.c        |   18 ++++++++++++++++++
 perliol.h       |    1 +
 pod/perliol.pod |   11 ++++++-----
 3 files changed, 25 insertions(+), 5 deletions(-)

diff --git a/perlio.c b/perlio.c
index 0eee430..85cb106 100644
--- a/perlio.c
+++ b/perlio.c
@@ -1315,6 +1315,24 @@ PerlIO_push(pTHX_ PerlIO *f, PERLIO_FUNCS_DECL(*tab), const char *mode, SV *arg)
     return f;
 }
 
+PerlIO *
+PerlIOBase_open(pTHX_ PerlIO_funcs *self, PerlIO_list_t *layers,
+	       IV n, const char *mode, int fd, int imode, int perm,
+	       PerlIO *old, int narg, SV **args)
+{
+    PerlIO_funcs * const tab = PerlIO_layer_fetch(aTHX_ layers, n - 1, PerlIO_default_layer(aTHX_ 0));
+    if (tab && tab->Open) {
+	PerlIO* ret = (*tab->Open)(aTHX_ tab, layers, n - 1, mode, fd, imode, perm, old, narg, args);
+	if (ret && PerlIO_push(aTHX_ ret, self, mode, PerlIOArg) == -1) {
+	    PerlIO_close(ret);
+	    return NULL;
+	}
+	return ret;
+    }
+    SETERRNO(EINVAL, LIB_INVARG);
+    return NULL;
+}
+
 IV
 PerlIOBase_binmode(pTHX_ PerlIO *f)
 {
diff --git a/perliol.h b/perliol.h
index 019fa8c..965910b 100644
--- a/perliol.h
+++ b/perliol.h
@@ -189,6 +189,7 @@ PERL_EXPORT_C IV        PerlIOBase_noop_fail(pTHX_ PerlIO *f);
 PERL_EXPORT_C IV        PerlIOBase_noop_ok(pTHX_ PerlIO *f);
 PERL_EXPORT_C IV        PerlIOBase_popped(pTHX_ PerlIO *f);
 PERL_EXPORT_C IV        PerlIOBase_pushed(pTHX_ PerlIO *f, const char *mode, SV *arg, PerlIO_funcs *tab);
+PERL_EXPORT_C PerlIO *  PerlIOBase_open(pTHX_ PerlIO_funcs *self, PerlIO_list_t *layers, IV n, const char *mode, int fd, int imode, int perm, PerlIO *old, int narg, SV **args);
 PERL_EXPORT_C SSize_t   PerlIOBase_read(pTHX_ PerlIO *f, void *vbuf, Size_t count);
 PERL_EXPORT_C void      PerlIOBase_setlinebuf(pTHX_ PerlIO *f);
 PERL_EXPORT_C SSize_t   PerlIOBase_unread(pTHX_ PerlIO *f, const void *vbuf, Size_t count);
diff --git a/pod/perliol.pod b/pod/perliol.pod
index e814847..0f71b93 100644
--- a/pod/perliol.pod
+++ b/pod/perliol.pod
@@ -526,11 +526,12 @@ passed to C<open>, otherwise it will be 1 if for example
 C<PerlIO_open> was called.  In simple cases SvPV_nolen(*args) is the
 pathname to open.
 
-Having said all that translation-only layers do not need to provide
-C<Open()> at all, but rather leave the opening to a lower level layer
-and wait to be "pushed".  If a layer does provide C<Open()> it should
-normally call the C<Open()> method of next layer down (if any) and
-then push itself on top if that succeeds.
+If a layer provides C<Open()> it should normally call the C<Open()>
+method of next layer down (if any) and then push itself on top if that
+succeeds.  C<PerlIOBase_open> is provided to do exactly that, so in
+most cases you don't have to write your own C<Open()> method.  If this
+method is not defined, other layers may have difficulty pushing
+themselves on top of it during open.
 
 If C<PerlIO_push> was performed and open has failed, it must
 C<PerlIO_pop> itself, since if it's not, the layer won't be removed
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>0002-Use-PerlIOBase_open-for-pop-utf8-and-bytes-layers.patch</summary>

```diff
From c28f016f24d055a7703df3c5ce0ccba7f2db7b40 Mon Sep 17 00:00:00 2001
From: Leon Timmermans <fawaka@gmail.com>
Date: Thu, 20 Jan 2011 22:04:39 +0100
Subject: [PATCH 2/3] Use PerlIOBase_open for pop, utf8 and bytes layers

---
 perlio.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/perlio.c b/perlio.c
index 85cb106..a89bf8b 100644
--- a/perlio.c
+++ b/perlio.c
@@ -1160,7 +1160,7 @@ PERLIO_FUNCS_DECL(PerlIO_remove) = {
     PERLIO_K_DUMMY | PERLIO_K_UTF8,
     PerlIOPop_pushed,
     NULL,
-    NULL,
+    PerlIOBase_open,
     NULL,
     NULL,
     NULL,
@@ -1966,7 +1966,7 @@ PERLIO_FUNCS_DECL(PerlIO_utf8) = {
     PERLIO_K_DUMMY | PERLIO_K_UTF8,
     PerlIOUtf8_pushed,
     NULL,
-    NULL,
+    PerlIOBase_open,
     NULL,
     NULL,
     NULL,
@@ -1997,7 +1997,7 @@ PERLIO_FUNCS_DECL(PerlIO_byte) = {
     PERLIO_K_DUMMY,
     PerlIOUtf8_pushed,
     NULL,
-    NULL,
+    PerlIOBase_open,
     NULL,
     NULL,
     NULL,
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>0003-Fixes-raw-layer-for-perl-80764.patch</summary>

```diff
From d2a4b7401a971b51bdeec78420b050a54ef67343 Mon Sep 17 00:00:00 2001
From: Leon Timmermans <fawaka@gmail.com>
Date: Thu, 20 Jan 2011 22:06:38 +0100
Subject: [PATCH 3/3] Fixes 'raw' layer for perl#80764

Made a ':raw' open do what it advertises to do (first open the file,
then binmode it), instead of leaving off the top layer.
---
 perlio.c |   16 +---------------
 1 files changed, 1 insertions(+), 15 deletions(-)

diff --git a/perlio.c b/perlio.c
index a89bf8b..3ce31d1 100644
--- a/perlio.c
+++ b/perlio.c
@@ -2021,20 +2021,6 @@ PERLIO_FUNCS_DECL(PerlIO_byte) = {
     NULL,                       /* set_ptrcnt */
 };
 
-PerlIO *
-PerlIORaw_open(pTHX_ PerlIO_funcs *self, PerlIO_list_t *layers,
-	       IV n, const char *mode, int fd, int imode, int perm,
-	       PerlIO *old, int narg, SV **args)
-{
-    PerlIO_funcs * const tab = PerlIO_default_btm();
-    PERL_UNUSED_ARG(self);
-    if (tab && tab->Open)
-	 return (*tab->Open) (aTHX_ tab, layers, n - 1, mode, fd, imode, perm,
-			      old, narg, args);
-    SETERRNO(EINVAL, LIB_INVARG);
-    return NULL;
-}
-
 PERLIO_FUNCS_DECL(PerlIO_raw) = {
     sizeof(PerlIO_funcs),
     "raw",
@@ -2042,7 +2028,7 @@ PERLIO_FUNCS_DECL(PerlIO_raw) = {
     PERLIO_K_DUMMY,
     PerlIORaw_pushed,
     PerlIOBase_popped,
-    PerlIORaw_open,
+    PerlIOBase_open,
     NULL,
     NULL,
     NULL,
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jan 20 13&#8203;:50&#8203;:07 2011\, LeonT wrote&#8203;:
> The problem behind this can be summarized to this&#8203;: «during open\(\) a
> PerlIO layer that defines an Open method can not be pushed on top of
> another layer that doesn't define Open»\. There are two possible
> workarounds for this&#8203;:
> 
> a\) Implement Open in all layers that don't implement it now\.
> b\) Change the lookup mechanism in all existing Open methods\.
> 
> I decided to implement option a\, because it seem to be the least
> intrusive way to tackle this problem\. In the first patch\, I introduce
> PerlIOBase\_open\, a generic open method that does what 90%\+ of all
> modules want it to do \(open the layer beneath it\, and then push the
> current layer on top of it\)\. It also updates the documentation to
> advise people to use it\.
> 
> The second patch used this to fix this bug \(\#82484\)\. The third patch
> fixes \#80764 using the same Open method \(consider my previous patch
> for that bug retracted in favor of this one\)\. These two patches are
> quite trivial :\-\)\.
> 
> Leon

Thank you\. Applied as&#8203;:

81fe74fb3f93457df8e864c91546ac6b860657fc
c0888ace5ec8d378714cd95caba0b35057e6c3f1
ecfd064986aef03b9b7d8b56da83ae37ff435721

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @Leont

On Thu\, Jan 20\, 2011 at 10&#8203;:49 PM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
> The problem behind this can be summarized to this&#8203;: «during open\(\) a
> PerlIO layer that defines an Open method can not be pushed on top of
> another layer that doesn't define Open»\. There are two possible
> workarounds for this&#8203;:
>
> a\) Implement Open in all layers that don't implement it now\.
> b\) Change the lookup mechanism in all existing Open methods\.
>
> I decided to implement option a\, because it seem to be the least
> intrusive way to tackle this problem\. In the first patch\, I introduce
> PerlIOBase\_open\, a generic open method that does what 90%\+ of all
> modules want it to do \(open the layer beneath it\, and then push the
> current layer on top of it\)\. It also updates the documentation to
> advise people to use it\.
>
> The second patch used this to fix this bug \(\#82484\)\. The third patch
> fixes \#80764 using the same Open method \(consider my previous patch
> for that bug retracted in favor of this one\)\. These two patches are
> quite trivial :\-\)\.
>
> Leon

It seems my patch missed something important&#8203;: it didn't export the
PerlIOBase\_open\(\)\. Missed it since they're exported weirdly\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>0001-Export-PerlIOBase_open.patch</summary>

```diff
From b1aa5e4a58782d648c8ba744ca9097e2193cf62e Mon Sep 17 00:00:00 2001
From: Leon Timmermans <fawaka@gmail.com>
Date: Mon, 4 Jul 2011 13:05:39 +0200
Subject: [PATCH] Export PerlIOBase_open

---
 makedef.pl |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/makedef.pl b/makedef.pl
index 0994fd9..16d4f9a 100644
--- a/makedef.pl
+++ b/makedef.pl
@@ -555,6 +555,7 @@ my @layer_syms = qw(
 		    PerlIOBase_eof
 		    PerlIOBase_error
 		    PerlIOBase_fileno
+		    PerlIOBase_open
 		    PerlIOBase_noop_fail
 		    PerlIOBase_noop_ok
 		    PerlIOBase_popped
-- 
1.7.5.4


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Dec 19 16&#8203;:09&#8203;:28 2011\, LeonT wrote&#8203;:
> On Thu\, Jan 20\, 2011 at 10&#8203;:49 PM\, Leon Timmermans \<fawaka@&#8203;gmail\.com>
wrote&#8203;:
> > The problem behind this can be summarized to this&#8203;: �during open\(\) a
> > PerlIO layer that defines an Open method can not be pushed on top of
> > another layer that doesn't define Open�\. There are two possible
> > workarounds for this&#8203;:
> >
> > a\) Implement Open in all layers that don't implement it now\.
> > b\) Change the lookup mechanism in all existing Open methods\.
> >
> > I decided to implement option a\, because it seem to be the least
> > intrusive way to tackle this problem\. In the first patch\, I introduce
> > PerlIOBase\_open\, a generic open method that does what 90%\+ of all
> > modules want it to do \(open the layer beneath it\, and then push the
> > current layer on top of it\)\. It also updates the documentation to
> > advise people to use it\.
> >
> > The second patch used this to fix this bug \(\#82484\)\. The third patch
> > fixes \#80764 using the same Open method \(consider my previous patch
> > for that bug retracted in favor of this one\)\. These two patches are
> > quite trivial :\-\)\.
> >
> > Leon
> 
> It seems my patch missed something important&#8203;: it didn't export the
> PerlIOBase\_open\(\)\. Missed it since they're exported weirdly\.
> 
> Leon

Thank you\.  Applied as 3dfd1b5cd2\.


\-\- 

Father Chrysostomos
